File: go/planet_direct_test.go
package voxgigsolardemosdk

import (
	"strings"
	"testing"
)

func TestPlanetDirect(t *testing.T) {
	t.Run("direct-load-planet", func(t *testing.T) {
		setup := planetDirectSetup(map[string]any{"id": "direct01"})
		client := setup.client
		calls := setup.calls

		result, err := client.Direct(map[string]any{
			"path":   "api/planet/{id}",
			"method": "GET",
			"params": map[string]any{"id": "direct01"},
		})
		if err != nil {
			t.Fatalf("direct failed: %v", err)
		}

		if result["ok"] != true {
			t.Fatalf("expected ok to be true, got %v", result["ok"])
		}
		if toInt(result["status"]) != 200 {
			t.Fatalf("expected status 200, got %v", result["status"])
		}
		if result["data"] == nil {
			t.Fatal("expected data to be non-nil")
		}
		if dataMap, ok := result["data"].(map[string]any); ok {
			if dataMap["id"] != "direct01" {
				t.Fatalf("expected data.id to be direct01, got %v", dataMap["id"])
			}
		}

		if len(*calls) != 1 {
			t.Fatalf("expected 1 call, got %d", len(*calls))
		}
		call := (*calls)[0]
		if initMap, ok := call["init"].(map[string]any); ok {
			if initMap["method"] != "GET" {
				t.Fatalf("expected method GET, got %v", initMap["method"])
			}
		}
		if url, ok := call["url"].(string); ok {
			if !strings.Contains(url, "direct01") {
				t.Fatalf("expected url to contain direct01, got %v", url)
			}
		}
	})

	t.Run("direct-list-planet", func(t *testing.T) {
		setup := planetDirectSetup([]any{
			map[string]any{"id": "direct01"},
			map[string]any{"id": "direct02"},
		})
		client := setup.client
		calls := setup.calls

		result, err := client.Direct(map[string]any{
			"path":   "api/planet",
			"method": "GET",
			"params": map[string]any{},
		})
		if err != nil {
			t.Fatalf("direct failed: %v", err)
		}

		if result["ok"] != true {
			t.Fatalf("expected ok to be true, got %v", result["ok"])
		}
		if toInt(result["status"]) != 200 {
			t.Fatalf("expected status 200, got %v", result["status"])
		}
		if dataList, ok := result["data"].([]any); ok {
			if len(dataList) != 2 {
				t.Fatalf("expected 2 items, got %d", len(dataList))
			}
		} else {
			t.Fatalf("expected data to be an array, got %T", result["data"])
		}

		if len(*calls) != 1 {
			t.Fatalf("expected 1 call, got %d", len(*calls))
		}
	})
}

type planetDirectSetupResult struct {
	client *SolardemoSDK
	calls  *[]map[string]any
}

func planetDirectSetup(mockres any) *planetDirectSetupResult {
	calls := &[]map[string]any{}

	mockFetch := func(url string, init map[string]any) (map[string]any, error) {
		*calls = append(*calls, map[string]any{"url": url, "init": init})
		return map[string]any{
			"status":     200,
			"statusText": "OK",
			"headers":    map[string]any{},
			"json": (func() any)(func() any {
				if mockres != nil {
					return mockres
				}
				return map[string]any{"id": "direct01"}
			}),
		}, nil
	}

	client := NewSolardemoSDK(map[string]any{
		"base": "http://localhost:8080",
		"system": map[string]any{
			"fetch": (func(string, map[string]any) (map[string]any, error))(mockFetch),
		},
	})

	return &planetDirectSetupResult{client: client, calls: calls}
}

File: scala/sdktest/PlanetDirectTest.scala
// Generated direct-call tests for the planet entity (mirrors the java
// TestDirect generator). A dependency-free scala-cli test object driven by
// SdkEntityTestMain: an offline mock transport records each call and the
// asserts confirm path-param substitution and the response shape.

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}
import java.util.function.{BiFunction, Supplier}

import voxgig.solardemosdk.core.{Helpers, SdkEntity, SolardemoSDK}

object PlanetDirectTest {

  private class DirectSetup(val client: SolardemoSDK, val calls: JList[JMap[String, Object]])

  private def directSetup(mockres: Object): DirectSetup = {
    val calls = new ArrayList[JMap[String, Object]]()
    val mockdata: Object = if (mockres != null) mockres else SdkTestSupport.om("id" -> "direct01")
    val mockFetch: BiFunction[String, JMap[String, Object], Object] =
      (url, init) => {
        calls.add(SdkTestSupport.om("url" -> url, "init" -> init))
        val js: Supplier[Object] = () => mockdata
        SdkTestSupport.om(
          "status" -> SdkTestSupport.I(200),
          "statusText" -> "OK",
          "headers" -> new LinkedHashMap[String, Object](),
          "json" -> js)
      }
    val client = new SolardemoSDK(SdkTestSupport.om(
      "base" -> "http://localhost:8080",
      "system" -> SdkTestSupport.om("fetch" -> mockFetch)))
    new DirectSetup(client, calls)
  }

  def run(rep: SdkTestReport): Unit = {
    rep.scope("direct-list-planet") {
      val setup = directSetup(SdkTestSupport.jl(
          SdkTestSupport.om("id" -> "direct01"),
          SdkTestSupport.om("id" -> "direct02")))
      val client = setup.client

      val params = new LinkedHashMap[String, Object]()
      val result = client.direct(SdkTestSupport.om(
          "path" -> "api/planet",
          "method" -> "GET",
          "params" -> params))

      rep.eq("direct-list-planet.ok", java.lang.Boolean.TRUE, result.get("ok"))
      rep.eqI("direct-list-planet.status", 200, Helpers.toInt(result.get("status")))
      rep.check("direct-list-planet.islist", result.get("data").isInstanceOf[JList[?]], "expected data to be an array, got " + result.get("data"))
      val listData = result.get("data").asInstanceOf[JList[Object]]
      rep.eqI("direct-list-planet.size", 2, listData.size())
      rep.eqI("direct-list-planet.calls", 1, setup.calls.size())
    }

    rep.scope("direct-load-planet") {
      val setup = directSetup(SdkTestSupport.om("id" -> "direct01"))
      val client = setup.client

      val params = new LinkedHashMap[String, Object]()
      params.put("id", "direct01")
      val result = client.direct(SdkTestSupport.om(
          "path" -> "api/planet/{id}",
          "method" -> "GET",
          "params" -> params))

      rep.eq("direct-load-planet.ok", java.lang.Boolean.TRUE, result.get("ok"))
      rep.eqI("direct-load-planet.status", 200, Helpers.toInt(result.get("status")))
      rep.check("direct-load-planet.data", result.get("data") != null, "expected data to be non-null")
      val dataMap = Helpers.toMapAny(result.get("data"))
      if (dataMap != null) rep.eq("direct-load-planet.dataId", "direct01", dataMap.get("id"))
      rep.eqI("direct-load-planet.calls", 1, setup.calls.size())
      val url = setup.calls.get(0).get("url") match { case s: String => s; case _ => "" }
      rep.check("direct-load-planet.url1", url.contains("direct01"), "expected url to contain direct01, got " + url)
    }
  }
}
